ngraph/core/src/op/einsum.cpp

```diff
@@ -0,0 +1,277 @@
+#include <cctype>
+#include <ngraph/validation_util.hpp>
+#include <regex>
+#include <string>
+#include <unordered_map>
+
+#include "itt.hpp"
+#include "ngraph/op/einsum.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+NGRAPH_RTTI_DEFINITION(op::v7::Einsum, "Einsum", 7);
+
+op::v7::Einsum::Einsum(const OutputVector& inputs, const std::string& equation)
+    : Op(inputs)
+    , m_equation(equation)
+{
+    constructor_validate_and_infer_types();
+}
+
+bool op::v7::Einsum::parse_equation(std::vector<std::string>& input_subscripts,
+                                    std::string& output_subscript)
+{
+    // remove extra white-spaces from the equation
+    m_equation.erase(std::remove_if(m_equation.begin(), m_equation.end(), ::isspace),
+                     m_equation.end());
+
+    // split equation to input subscripts and an output subscript
+    auto pos_output_delimeter = m_equation.find("->");
+    auto input_subscripts_str = m_equation.substr(0, pos_output_delimeter);
+
+    // split the input subscripts into a vector of input subscripts
+    input_subscripts.clear();
+    std::istringstream input;
+    input.str(input_subscripts_str);
+    for (std::string input_subscript; std::getline(input, input_subscript, ',');)
+    {
+        input_subscripts.push_back(input_subscript);
+    }
+
+    // check that each input subscript contains only alphabetic letter or ellipsis
+    const std::regex subscript_regex("[A-Za-z]*|[A-Za-z]*\\.\\.\\.[A-Za-z]*");
+    for (auto const& input_subscript : input_subscripts)
+    {
+        if (std::regex_match(input_subscript, subscript_regex) == false)
+        {
+            return false;
+        }
+    }
+
+    // figure out if ellipsis is met in input subscripts
+    const std::regex ellipsis_regex("\\.\\.\\.");
+    bool is_ellipsis_met = false;
+    std::smatch ellipsis_match;
+    for (auto const& input_subscript : input_subscripts)
+    {
+        if (std::regex_search(input_subscript, ellipsis_match, ellipsis_regex) == true)
+        {
+            is_ellipsis_met = true;
+            break;
+        }
+    }
+
+    if (pos_output_delimeter == std::string::npos)
+    {
+        // recover output subscript
+        output_subscript = "";
+        for (auto const& input_subscript : input_subscripts)
+        {
+            for (auto const& label : input_subscript)
+            {
+                if (std::isalpha(label) && output_subscript.find(label) == std::string::npos)
+                {
+                    output_subscript += label;
+                }
+            }
+        }
+        std::sort(output_subscript.begin(), output_subscript.end());
+        if (is_ellipsis_met)
+        {
+            output_subscript = "..." + output_subscript;
+        }
+    }
+    else
+    {
+        output_subscript = m_equation.substr(pos_output_delimeter + 2);
+
+        // check that the output subscript has the correct format
+        if (std::regex_match(output_subscript, subscript_regex) == false)
```

Comment:
if (!std::regex_match(output_subscript, subscript_regex))

Reply:
The same comment as above.